examples/nanoChatGPT/train_reward.py

```diff
@@ -0,0 +1,412 @@
+
+    # FIXME: which one?
+    # optimizer = torch.optim.AdamW(model.model.reward_head.parameters(), lr=1e-3)
+    optimizer = torch.optim.AdamW(model.model.parameters(), lr=1e-4)
+
+    # wrap model into DDP container
+    if config["is_ddp"]:
+        model = DDP(model, device_ids=[config["ddp_local_rank"]])
+
+    # training loop
+    t0 = time.time()
+    local_iter_num = 0  # number of iterations in the lifetime of this process
+    config["running_mfu"] = -1.0
+    raw_model = (
+        model.module.module if config["is_ddp"] else model.module
+    )  # unwrap DDP container if needed
+    loss = None
+    # these will already have been set if resuming from previous checkpoint
+    iter_num = config.setdefault("iter_num", 0)
+    best_val_loss = config.setdefault("best_val_loss", 1e9)
+
+    estimate_loss = create_loss_estimator(config)
+
+    if config["decay_lr"]:
+        lr_scheduler = create_lr_scheduler(config)
+    else:
+
+        def lr_scheduler(_):
+            return config["learning_rate"]
+
+    while True:
+        # determine and set the learning rate for this iteration
+        lr = lr_scheduler(iter_num)
+
+        for param_group in optimizer.param_groups:
+            param_group["lr"] = lr
+
+        # # every once in a while evaluate the loss on train and val sets
+        if iter_num % config["eval_interval"] == 0 and config["master_process"]:
+            model.eval()
+            losses = {
+                "train": estimate_loss(model, train_loader),
+                "val": estimate_loss(model, val_loader),
+            }
+            model.train()
+            print(
+                f"step {iter_num}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}"
+            )
+            if losses["val"] < best_val_loss or config["always_save_checkpoint"]:
+                best_val_loss = losses["val"]
+                if iter_num > 0:
+                    checkpoint = {
+                        "model": raw_model.state_dict(),
+                        "optimizer": optimizer.state_dict(),
+                        "model_kwargs": model_kwargs,
+                        "iter_num": iter_num,
+                        "best_val_loss": best_val_loss,
+                        "config": config,
+                    }
+                    print(f"saving checkpoint to {config['out_dir']}")
+                    torch.save(checkpoint, os.path.join(config["out_dir"], "ckpt.pt"))
+        if iter_num == 0 and config["eval_only"]:
+            break
+
+        batch = next(train_loader)
+
+        # TODO: check why is different from std model (missing micro gradients)
+        
+        # TODO: combine evaluate_loss function with this. it's almost the same thing
+        # evaluate the loss
+        reward_chosen = model(batch.chosen)
+        reward_rejected = model(batch.rejected)
+        loss = -torch.log(torch.sigmoid(reward_chosen.reward - reward_rejected.reward)).mean()
```

Comment:
Since we wrapped model with `TensorDictModule` above, we should be passing in the tensorclass here, rather than a tensor.

But also the model expects the attribute `input` so we need to do something like

```python
batch_chosen = batch.select("chosen", "reward").rename_key_("chosen", "input")
batch_rejected = batch.select("rejected", "reward").rename_key_("rejected", "input")

model(batch_chosen)
model(batch_rejected)
```

Then we can rewrite `loss` as 

```python
loss = -torch.log(torch.sigmoid(batch_chosen.reward - batch_rejected.reward)).mean()
```

Reply:
This does feel a little bit clunky though

Reply:
the code above is not working as reward_chosen and reward_rejected are a tensor.
Instead of creating the tensordict we can just remove the ".reward" from the loss.